feat: Access log

gavindsouza opened the pull request:
for logging information about private file downloads, backups downloads, printing and exporting data

gavindsouza committed:
feat: Access log doctype for logging information about private file
downloads, backups downloads, printing and exporting data

gavindsouza committed:
updated: access log

gavindsouza committed:
updated: formattted access_log.js

gavindsouza committed:
access logging: print/pdf/pvt files/export/backup

client side implementation: treeview, query_report
server side implementation on the rest

gavindsouza committed:
fixes: codacy errors

gavindsouza committed:
Merge branch 'develop' into access-log

gavindsouza committed:
Update response.py

gavindsouza committed:
Update response.py

gavindsouza committed:
Update response.py

gavindsouza committed:
codacy: please

gavindsouza committed:
travis fixes

gavindsouza committed:
Merge branch 'access-log' of github.com:gavindsouza/frappe into access-log

surajshetty3416 commented on frappe/core/doctype/access_log/access_log.py:
```diff
@@ -0,0 +1,34 @@
+
+    frappe.get_doc({
+        'doctype': 'Access Log',
+        'name': "AL-{}-{}".format(user_name, frappe.generate_hash(length=5)),
```
Use in-built auto name feature instead...
<img width="455" alt="Screenshot 2019-07-30 at 10 51 56 PM" src="https://user-images.githubusercontent.com/13928957/62150880-bb0a9580-b31c-11e9-9720-4575febb2e64.png">

surajshetty3416 commented on frappe/core/doctype/access_log/access_log.py:
```diff
@@ -0,0 +1,34 @@
+
+@frappe.whitelist()
+def make_access_log(doctype=None, method=None, file_type=None, document=None, filters=None, report_name=None, page=None, backup=False):
+    user = frappe.get_user().load_user()
```
```suggestion
    user = frappe.session.user
```

surajshetty3416 commented on frappe/core/doctype/access_log/access_log.py:
```diff
@@ -0,0 +1,34 @@
+
+
+class AccessLog(Document):
+    pass
```
use tabs for indentation

https://github.com/frappe/erpnext/wiki/Coding-Standards#tabs-not-spaces

surajshetty3416 commented on frappe/core/doctype/access_log/access_log.py:
```diff
@@ -0,0 +1,34 @@
+    frappe.get_doc({
+        'doctype': 'Access Log',
+        'name': "AL-{}-{}".format(user_name, frappe.generate_hash(length=5)),
+        'user': user.first_name,
```
Since `user` is a link field, its value should be the primary key of the user.

```suggestion
        'user': frappe.session.user, # or "user" from above
```

surajshetty3416 commented on frappe/core/doctype/access_log/access_log.py:
```diff
@@ -0,0 +1,34 @@
+        'doctype': 'Access Log',
+        'name': "AL-{}-{}".format(user_name, frappe.generate_hash(length=5)),
+        'user': user.first_name,
+        'export_from': frappe.get_doc('DocType', doctype),
```
why?
```suggestion
        'export_from': doctype,
```

gavindsouza replied:
@scmmishra was helping me resolve some travis errors at the time

surajshetty3416 commented on frappe/core/doctype/access_log/access_log.py:
```diff
@@ -0,0 +1,34 @@
+        'method': method,
+        'filters': filters
+    }).insert()
+    frappe.db.commit()
```
Not required...
```suggestion
```

gavindsouza replied:
in case of GET requests, documents don't get inserted without a `frappe.db.commit`

surajshetty3416 commented on frappe/core/doctype/access_log/access_log.py:
```diff
@@ -0,0 +1,34 @@
+        'page': page,
+        'method': method,
+        'filters': filters
+    }).insert()
```
All users may not have access to this DocType
```suggestion
    }).insert(ignore_permissions=True)
```

surajshetty3416 commented on frappe/core/doctype/access_log/access_log.py:
```diff
@@ -0,0 +1,34 @@
+@frappe.whitelist()
+def make_access_log(doctype=None, method=None, file_type=None, document=None, filters=None, report_name=None, page=None, backup=False):
+    user = frappe.get_user().load_user()
+    user_name = user.username if user.username else user.first_name
```
```suggestion

```

surajshetty3416 commented on frappe/core/doctype/access_log/access_log.py:
```diff
@@ -0,0 +1,34 @@
+    user = frappe.get_user().load_user()
+    user_name = user.username if user.username else user.first_name
+
+    if backup:
```
Why do we need a special flag for backup? Can pass `report_name = 'Backup'` directly...

surajshetty3416 commented on frappe/core/doctype/access_log/access_log.json:
```diff
@@ -0,0 +1,138 @@
+  },
+  {
+   "default": "Now",
+   "fieldname": "timestamp",
```
seems to be a redundant field

We already have `creation`

gavindsouza replied:
should i display it from creation in the Form by updating an empty field?

gavindsouza committed:
changes:filters=>_filters(lol this works @ times?)

added: tests for access_log, make_access_log formatting, form ui added
buttons

gavindsouza committed:
this hopefully fixes em tests

gavindsouza committed:
changes: fixed indents, session.user, formatting

gavindsouza committed:
fixes: tests, refactor, permissions

emerging from exporter.py => reference doctype in export_data (check
format!)

gavindsouza committed:
fixes: convert dict to json

surajshetty3416 commented on frappe/core/doctype/access_log/access_log.js:
```diff
@@ -0,0 +1,26 @@
+
+frappe.ui.form.on('Access Log', {
+	refresh: function (frm) {
+		if (frm.doc.hasOwnProperty('report_name') && frm.doc.report_name != 'Backup') {
```
Use standard 'Display Depends On' feature for this...

surajshetty3416 commented on frappe/core/doctype/access_log/access_log.js:
```diff
@@ -0,0 +1,26 @@
+			try {
+				frappe.set_route('query-report', frm.doc.report_name, frm.doc.filters ? JSON.parse(frm.doc.filters) : '');
+			} catch (err) {
+				frappe.throw(__(err + ' has occurred'));
```
Showing such error message might not be useful for the end-user because the user cannot take any action using it.

Try to avoid the error you expect that might occur instead.

surajshetty3416 commented on frappe/core/doctype/access_log/test_access_log.py:
```diff
@@ -0,0 +1,108 @@
+
+
+class TestAccessLog(unittest.TestCase):
+	def test_make_full_access_log(self):
```
@gavindsouza

You might have to change the test case. Currently, you are testing if the method is able to create a document or not which is not that useful.

You can write a test to check if the access log gets created when you try to export data...

To do this, you can manually call `export_data` method and check if the access log has been created with the right required information.

Name of the test case can be `test_access_log_creation_for_data_export`

Similarly, you can write a test case for private file access as well...

gavindsouza replied:
tests for private file access and exporting csv done

gavindsouza committed:
changes: added more tests, wrote lesser code?

switched js to display depends on

gavindsouza committed:
fixes: codacy and hopefully travis

surajshetty3416 commented on frappe/core/doctype/access_log/access_log.js:
```diff
@@ -0,0 +1,17 @@
+		if (frm.doc.report_name.includes('/')) {
+			frappe.set_route(frm.doc.report_name);
+		} else {
+			let filters = frm.doc.filters ? frm.doc.filters : {};
```
Are you sure this works? I think you need to do `JSON.parse` because `frm.doc.filters` is saved as a JSON string.

gavindsouza replied:
writing tests makes a man do crazy things sometimes (like becoming a caffeine addict and forgetting to hard reload). 

Thanks for catching that!

surajshetty3416 commented:
@frappe-bot create instance

gavindsouza committed:
fix: tests? and polishing up

surajshetty3416 commented on frappe/core/doctype/access_log/access_log.json:
```diff
@@ -0,0 +1,144 @@
+ "sort_field": "modified",
+ "sort_order": "DESC",
+ "track_seen": 1,
+ "track_views": 1
```
Not required.
<img width="1200" alt="Screenshot 2019-08-05 at 3 51 52 PM" src="https://user-images.githubusercontent.com/13928957/62458842-9179c000-b79b-11e9-9d18-42a3142f0ba8.png">

gavindsouza replied:
permissions too? or just the track_views?

surajshetty3416 replied:
Just `track_views`

surajshetty3416 reviewed (changes requested):
@gavindsouza make access log for the access of [this method](https://github.com/frappe/frappe/blob/develop/frappe/desk/reportview.py#L117) as well (it gets called while exporting a report from list view)

--------------------
Also,
<img width="947" alt="Screenshot 2019-08-05 at 3 54 28 PM" src="https://user-images.githubusercontent.com/13928957/62458914-b66e3300-b79b-11e9-9431-37d4d1d4f640.png">
is this output expected for HTML page?

gavindsouza commented:
> <img alt="Screenshot 2019-08-05 at 3 54 28 PM" width="947" src="https://user-images.githubusercontent.com/13928957/62458914-b66e3300-b79b-11e9-9431-37d4d1d4f640.png">
> is this output expected for HTML page?

@surajshetty3416 the 'HTML page' output is expects columns selected when ```export_csv``` is called, the only case in which it's used to store this. Should that be another field instead, or perhaps a change in the fieldname is in order?

gavindsouza commented:
> @gavindsouza make access log for the access of [this method](https://github.com/frappe/frappe/blob/develop/frappe/desk/reportview.py#L117) as well (it gets called while exporting a report from list view)

on it 👍

gavindsouza committed:
added column field, removed tracked views

changed access log in exporter.py > export_data
added access log in reportview.py > export_query

surajshetty3416 committed:
style: Fix indentation

surajshetty3416 committed:
style: Fix indentation

surajshetty3416 committed:
fix: Remove problematic string replace

It may replace quotes in the actual filter values which can result in invalid JSON

surajshetty3416 committed:
fix: Remove filter string conversion

surajshetty3416 committed:
fix: Error due to quote escapes

TODO: Check for a better way
